pkg/objectfile/object_file.go

```diff
@@ -0,0 +1,270 @@
+		}
+	}
+
+	var (
+		kernelOffset *uint64
+		pageAligned  = func(addr uint64) bool { return addr%4096 == 0 }
+	)
+	if strings.Contains(filePath, "vmlinux") || !pageAligned(start) || !pageAligned(limit) || !pageAligned(offset) {
+		// Reading all Symbols is expensive, and we only rarely need it so
+		// we don't want to do it every time. But if _stext happens to be
+		// page-aligned but isn't the same as Vaddr, we would symbolize
+		// wrong. So if the name the addresses aren't page aligned, or if
+		// the name is "vmlinux" we read _stext. We can be wrong if: (1)
+		// someone passes a kernel path that doesn't contain "vmlinux" AND
+		// (2) _stext is page-aligned AND (3) _stext is not at Vaddr
+		symbols, err := ef.Symbols()
+		if err != nil && err != elf.ErrNoSymbols {
+			return nil, err
+		}
+
+		// The kernel relocation symbol (the mapping start address) can be either
+		// _text or _stext. When profiles are generated by `perf`, which one was used is
+		// distinguished by the mapping name for the kernel image:
+		// '[kernel.kallsyms]_text' or '[kernel.kallsyms]_stext', respectively. If we haven't
+		// been able to parse it from the mapping, we default to _stext.
+		if relocationSymbol == "" {
+			relocationSymbol = "_stext"
+		}
+		for _, s := range symbols {
+			if s.Name == relocationSymbol {
+				kernelOffset = &s.Value
+				break
+			}
+		}
+	}
+
+	// Check that we can compute a base for the binary. This may not be the
+	// correct base value, so we don't save it. We delay computing the actual base
+	// value until we have a sample address for this mapping, so that we can
+	// correctly identify the associated program segment that is needed to compute
+	// the base.
+	if _, err := elfexec.GetBase(&ef.FileHeader, elfexec.FindTextProgHeader(ef), kernelOffset, start, limit, offset); err != nil {
+		return nil, fmt.Errorf("could not identify base for %s: %v", filePath, err)
+	}
+	return &ObjectFile{
+		BuildID: buildID,
+		path:    filePath,
+		m: &mapping{
+			start:        start,
+			limit:        limit,
+			offset:       offset,
+			kernelOffset: kernelOffset,
+		},
+	}, nil
+}
+
+func NewObjectFile(pid uint32, m *profile.Mapping) *ObjectFile {
+	return &ObjectFile{
+		PID:     pid,
+		File:    m.File,
+		BuildID: m.BuildID,
+		m: &mapping{
+			start:  m.Start,
+			limit:  m.Limit,
+			offset: m.Offset,
+		},
+	}
+}
+
+type ObjectFile struct {
+	PID     uint32
+	File    string
+	BuildID string
+
+	path string
+
+	// Ensures the base, baseErr and isData are computed once.
+	baseOnce sync.Once
+	base     uint64
+	baseErr  error
+
+	isData bool
+	m      *mapping
+}
+
+func (f *ObjectFile) Root() string {
+	return path.Join("/proc", strconv.FormatUint(uint64(f.PID), 10), "/root")
+}
+
+func (f *ObjectFile) FullPath() string {
+	if f.path != "" {
+		return f.path
+	}
+	return path.Join(f.Root(), f.File)
+}
+
+func (f *ObjectFile) ObjAddr(addr uint64) (uint64, error) {
+	f.baseOnce.Do(func() { f.baseErr = f.computeBase(addr) })
```

Comment:
we already have the `NewObjectFile` function that we use to instantiate the struct, let's just do the computation there once

Reply:
In order to compute the base properly we need an address, that's why we need to defer the calculation until we received an address from the profiler.